internal/command/user_v2_human.go

```diff
@@ -0,0 +1,559 @@
+		if err != nil {
+			return err
+		}
+	}
+	if human.Email != nil {
+		cmds, human.EmailCode, err = c.changeUserEmail(ctx, cmds, existingHuman, human.Email, alg)
+		if err != nil {
+			return err
+		}
+	}
+	if human.Phone != nil {
+		cmds, human.PhoneCode, err = c.changeUserPhone(ctx, cmds, existingHuman, human.Phone, alg)
+		if err != nil {
+			return err
+		}
+	}
+	if human.Password != nil {
+		cmds, err = c.changeUserPassword(ctx, cmds, existingHuman, human.Password, alg)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(cmds) == 0 {
+		human.Details = writeModelToObjectDetails(&existingHuman.WriteModel)
+		return nil
+	}
+	err = c.pushAppendAndReduce(ctx, existingHuman, cmds...)
+	if err != nil {
+		return err
+	}
+	human.Details = writeModelToObjectDetails(&existingHuman.WriteModel)
+	return nil
+}
+
+func (c *Commands) changeUserEmail(ctx context.Context, cmds []eventstore.Command, wm *UserV2WriteModel, email *Email, alg crypto.EncryptionAlgorithm) (_ []eventstore.Command, code *string, err error) {
+	ctx, span := tracing.NewSpan(ctx)
+	defer func() { span.End() }()
+
+	if email.Address != "" {
+		cmd := wm.NewEmailAddressChangedEvent(ctx, email.Address)
+		if cmd != nil {
+			cmds = append(cmds, cmd)
+		}
+	}
+	if email.Verified {
+		cmd := wm.NewEmailIsVerifiedEvent(ctx, email.Verified)
+		if cmd != nil {
+			cmds = append(cmds, cmd)
+		}
+	} else {
+		// if email not changed or empty, send no code
+		if email.Address == "" || email.Address == wm.Email {
+			return cmds, code, err
+		}
+		c, err := c.newEmailCode(ctx, c.eventstore.Filter, alg) //nolint:staticcheck
+		if err != nil {
+			return cmds, code, err
+		}
+		cmds = append(cmds, user.NewHumanEmailCodeAddedEventV2(ctx, &wm.Aggregate().Aggregate, c.Crypted, c.Expiry, email.URLTemplate, email.ReturnCode))
+		if email.ReturnCode {
+			code = &c.Plain
+		}
+	}
+	return cmds, code, nil
+}
+
+func (c *Commands) changeUserPhone(ctx context.Context, cmds []eventstore.Command, wm *UserV2WriteModel, phone *Phone, alg crypto.EncryptionAlgorithm) (_ []eventstore.Command, code *string, err error) {
+	ctx, span := tracing.NewSpan(ctx)
+	defer func() { span.End() }()
+
+	if phone.Number != "" {
+		cmd := wm.NewPhoneNumberChangedEvent(ctx, phone.Number)
+		if cmd != nil {
+			cmds = append(cmds, cmd)
+		}
+	}
+	if phone.Verified {
+		cmd := wm.NewPhoneIsVerifiedEvent(ctx, phone.Verified)
+		if cmd != nil {
+			cmds = append(cmds, cmd)
+		}
+	} else {
+		// if phone not changed or empty, send no code
+		if phone.Number == "" || phone.Number == wm.Phone {
+			return cmds, code, err
+		}
+		c, err := c.newPhoneCode(ctx, c.eventstore.Filter, alg) //nolint:staticcheck
+		if err != nil {
+			return cmds, code, err
+		}
+		cmds = append(cmds, user.NewHumanPhoneCodeAddedEventV2(ctx, &wm.Aggregate().Aggregate, c.Crypted, c.Expiry, phone.ReturnCode))
+		if phone.ReturnCode {
+			code = &c.Plain
+		}
+	}
+	return cmds, code, nil
+}
```

Comment:
IMO these functions are hard to read... lot of tiny bits and pieces...

e.g. why not just early return if phone is verified
also `wm.NewPhoneIsVerifiedEvent` is only called, if phone.Verified, which is also passed into the function, where if it would be false (how?) no event is returned. this then needs to be additionally checked by the caller 🤯

Reply:
Structured email and phone differently